File: sdk/envelope.nim
## Envelope codec — Niffler wire protocol (docs/WIRE.md)
##
## Pure std/json so the SDK stays trivially portable; this is runtime data
## by design (schemas-as-runtime-data), not a typed seam.

import std/[json, os, times]

type
  EnvelopeKind* = enum
    ekCall = "call"
    ekResult = "result"
    ekEvent = "event"
    ekError = "error"

  Envelope* = object
    v*: int
    id*: string
    kind*: EnvelopeKind
    tool*: string
    args*: JsonNode     ## call/result only
    payload*: JsonNode  ## event only
    error*: JsonNode    ## error only: {"code": "...", "message": "..."}

var idCounter = 0

proc newId*(): string =
  ## Cheap unique id (pid + ns timestamp + counter; plenty for a bus id)
  inc idCounter
  result = $(epochTime() * 1_000_000_000).int64 & "-" & $getCurrentProcessId() & "-" & $idCounter

proc parseKind(s: string): EnvelopeKind =
  case s
  of "call": ekCall
  of "result": ekResult
  of "event": ekEvent
  else: ekError

proc toJson*(e: Envelope): JsonNode =
  result = %*{"v": e.v, "id": e.id, "kind": $e.kind}
  if e.tool.len > 0: result["tool"] = %e.tool
  if e.args != nil: result["args"] = e.args
  if e.payload != nil: result["payload"] = e.payload
  if e.error != nil: result["error"] = e.error

proc fromJson*(node: JsonNode): Envelope =
  result.v = node{"v"}.getInt(1)
  result.id = node{"id"}.getStr("")
  result.kind = parseKind(node{"kind"}.getStr("event"))
  result.tool = node{"tool"}.getStr("")
  result.args = node{"args"}
  result.payload = node{"payload"}
  result.error = node{"error"}

proc encode*(e: Envelope): string = $e.toJson()

proc decode*(data: string): Envelope =
  try:
    return data.parseJson().fromJson()
  except CatchableError:
    # Never crash the component on garbage; report it as an error envelope
    return Envelope(v: 1, id: newId(), kind: ekError,
                    error: %*{"code": "bad-envelope",
                              "message": data[0 ..< min(200, data.len)]})

proc errorEnvelope*(id, code, message: string): Envelope =
  Envelope(v: 1, id: id, kind: ekError, error: %*{"code": code, "message": message})

proc resultEnvelope*(id: string, value: JsonNode): Envelope =
  Envelope(v: 1, id: id, kind: ekResult, args: value)

proc callEnvelope*(tool: string, args: JsonNode): Envelope =
  Envelope(v: 1, id: newId(), kind: ekCall, tool: tool, args: args)

File: sdk/ts/src/envelope.ts
// Niffler envelope codec — the wire protocol (docs/WIRE.md).
//
// Mirrors sdk/envelope.nim and sdk/go/envelope.go 1:1 — pure JSON runtime
// data, so it stays trivially portable. The envelope is the artifact.

export type EnvelopeKind = "call" | "result" | "event" | "error";

export interface ErrorInfo {
  code: string;
  message: string;
}

export interface Envelope {
  v: number;
  id: string;
  kind: EnvelopeKind;
  tool?: string;
  args?: unknown;
  payload?: unknown;
  error?: ErrorInfo;
  /** Component name of the call originator (call only; self-declared
   *  routing hint, not an auth claim). */
  caller?: string;
}

let idCounter = 0;

/** Cheap unique bus id (ns timestamp + pid + counter). */
export function newId(): string {
  idCounter++;
  return `${Date.now() * 1_000_000}-${process.pid}-${idCounter}`;
}

function parseKind(s: string): EnvelopeKind {
  switch (s) {
    case "call":
      return "call";
    case "result":
      return "result";
    case "event":
      return "event";
    default:
      return "error";
  }
}

export function toJson(e: Envelope): Record<string, unknown> {
  const out: Record<string, unknown> = { v: e.v, id: e.id, kind: e.kind };
  if (e.tool) out.tool = e.tool;
  if (e.args !== undefined) out.args = e.args;
  if (e.payload !== undefined) out.payload = e.payload;
  if (e.error !== undefined) out.error = e.error;
  if (e.caller) out.caller = e.caller;
  return out;
}

export function fromJson(node: Record<string, unknown>): Envelope {
  return {
    v: (node.v as number) ?? 1,
    id: (node.id as string) ?? "",
    kind: parseKind((node.kind as string) ?? "event"),
    tool: (node.tool as string) ?? "",
    args: node.args,
    payload: node.payload,
    error: node.error as ErrorInfo | undefined,
    caller: (node.caller as string) ?? "",
  };
}

export function encode(e: Envelope): string {
  return JSON.stringify(toJson(e));
}

/** Decode an envelope; malformed input becomes an error envelope, never a crash. */
export function decode(data: string): Envelope {
  try {
    return fromJson(JSON.parse(data) as Record<string, unknown>);
  } catch (err) {
    return {
      v: 1,
      id: newId(),
      kind: "error",
      error: { code: "bad-envelope", message: String(err) },
    };
  }
}

export function errorEnvelope(id: string, code: string, message: string): Envelope {
  return { v: 1, id, kind: "error", error: { code, message } };
}

export function resultEnvelope(id: string, value: unknown): Envelope {
  return { v: 1, id, kind: "result", args: value };
}

export function callEnvelope(tool: string, args: unknown): Envelope {
  return { v: 1, id: newId(), kind: "call", tool, args };
}
